From 6cf4814be6b971f0e7141f9bb9a3acd992d576bf Mon Sep 17 00:00:00 2001
Message-Id: <6cf4814be6b971f0e7141f9bb9a3acd992d576bf.1426664743.git.chang-joon.lee@intel.com>
From: vkorjani <vikas.korjani@intel.com>
Date: Wed, 4 Mar 2015 10:51:20 +0530
Subject: [PATCH 1/3] REVERTME [VPG]: drm/i915: Check FIFO !FULL before
 sending long packet

This patch we check FIFO !FULL instead of FIFO empty before sending
long packet.

REVERTME: This change is already available in upstream, reference patch
Upstream-Repo: http://anongit.freedesktop.org/git/drm-intel
commit 69c05eb2d8421cbe96c2c66ea6e3223d25928d92
Author: Jani Nikula <jani.nikula@intel.com>.
drm/i915: add MIPI DSI command sending routines
When above upstream patch is merged in next forklift, this patch can be reverted.

Issue: IRDA-3305
Change-Id: Iedfc55f0b2aeae0ddbf9b9af90a8bbc1d02faae9
Signed-off-by: vkorjani <vikas.korjani@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi_cmd.c |   17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dsi_cmd.c b/drivers/gpu/drm/i915/intel_dsi_cmd.c
index 92044ff..cc615ac 100644
--- a/drivers/gpu/drm/i915/intel_dsi_cmd.c
+++ b/drivers/gpu/drm/i915/intel_dsi_cmd.c
@@ -174,7 +174,7 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 	enum pipe pipe = intel_dsi->port;
 	u32 data_reg, ctrl_reg, ctrl;
 	int i, j, n;
-	u32 mask;
+	u32 ctrl_mask, data_mask;
 
 	DRM_DEBUG_KMS("channel %d, data_type %d, len %04x\n",
 		      channel, data_type, len);
@@ -182,15 +182,18 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 	if (intel_dsi->hs) {
 		data_reg = MIPI_HS_GEN_DATA(pipe);
 		ctrl_reg = MIPI_HS_GEN_CTRL(pipe);
-		mask = HS_CTRL_FIFO_EMPTY | HS_DATA_FIFO_EMPTY;
+		ctrl_mask = HS_CTRL_FIFO_FULL;
+		data_mask = HS_DATA_FIFO_FULL;
 	} else {
 		data_reg = MIPI_LP_GEN_DATA(pipe);
 		ctrl_reg = MIPI_LP_GEN_CTRL(pipe);
-		mask = LP_CTRL_FIFO_EMPTY | LP_DATA_FIFO_EMPTY;
+		ctrl_mask = LP_CTRL_FIFO_FULL;
+		data_mask = LP_DATA_FIFO_FULL;
 	}
 
-	if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask) == mask, 50))
-		DRM_ERROR("Timeout waiting for HS/LP DATA FIFO to be empty\n");
+	if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) &
+				data_mask) == 0, 50))
+		DRM_ERROR("Timeout for HS/LP DATA FIFO to be !FULL\n");
 
 	for (i = 0; i < len; i += n) {
 		u32 val = 0;
@@ -207,6 +210,10 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 	ctrl |= channel << VIRTUAL_CHANNEL_SHIFT;
 	ctrl |= data_type << DATA_TYPE_SHIFT;
 
+	if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) &
+					ctrl_mask) == 0, 50))
+		DRM_ERROR("Timeout for HS/LP CTRL FIFO to be !FULL\n");
+
 	I915_WRITE(ctrl_reg, ctrl);
 	return 0;
 
-- 
1.7.9.5

